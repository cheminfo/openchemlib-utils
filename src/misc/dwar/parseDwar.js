import { ensureString } from 'ensure-string';

import { getParts } from './utils/getParts';
import { parseColumnbProperties } from './utils/parseColumnbProperties';
import { parseData } from './utils/parseData';

/**
 * Convert a DataWarrior database into a JSON object
 * @param {string} text
 * @returns
 */
export function parseDwar(text) {
  text = ensureString(text);
  const parts = getParts(text);
  improveParts(parts);
  return parts;
}

export function getCamelCase(name) {
  return name.replace(/[ -][a-z]/g, (string) => string[1].toUpperCase());
}

function improveParts(parts) {
  for (const key in parts) {
    switch (key) {
      case 'columnProperties':
        parts[key] = parseColumnbProperties(parts[key]);
        break;
      case 'data':
        break;
      default:
        parts[key] = parseDefault(parts[key]);
    }
  }
  const data = parseData(parts.data, {
    columnProperties: parts.columnProperties,
  });
  parts.data = data.entries;
  parts.rawData = data.rawEntries;
}

function parseDefault(lines) {
  const result = {};
  for (const line of lines) {
    const [key, value] = line.slice(1, -1).split('=');
    result[key] = value.slice(1, -1);
  }
  return result;
}
