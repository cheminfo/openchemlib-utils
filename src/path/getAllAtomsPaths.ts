import { Molecule } from 'openchemlib';

export interface GetAllAtomsPathsOptions {
  /**
   * Maximum path length to consider. All the paths from 0 to maxPathLength will be considered
   * @default 5
   */
  maxPathLength?: number;
}

interface AtomPath {
  path: number[];
  distance: number;
}

/**
 * We need to create an array of atoms
 * that contains an array of pathLength
 * that contains an array of object
 * @param molecule
 * @param options
 * @returns
 */
export function getAllAtomsPaths(
  molecule: Molecule,
  options: GetAllAtomsPathsOptions = {},
) {
  const { maxPathLength = 5 } = options;
  const allAtomsPaths = [];
  for (let i = 0; i < molecule.getAllAtoms(); i++) {
    const oneAtomPaths: AtomPath[][] = [];
    allAtomsPaths.push(oneAtomPaths);

    let atomPaths: AtomPath[] = [];
    atomPaths.push({ path: [i], distance: 0 });
    oneAtomPaths.push(atomPaths);

    let nextIndexes = [0];
    let nextAtoms = [i];

    for (let sphere = 1; sphere <= maxPathLength; sphere++) {
      atomPaths = [];
      oneAtomPaths.push(atomPaths);
      const currentIndexes = nextIndexes;
      const currentAtoms = nextAtoms;
      nextIndexes = [];
      nextAtoms = [];

      for (let i = 0; i < currentIndexes.length; i++) {
        const atom = currentAtoms[i];
        const index = currentIndexes[i];

        const previousPath = oneAtomPaths[sphere - 1][index].path;
        for (let conn = 0; conn < molecule.getConnAtoms(atom); conn++) {
          const connectedAtom = molecule.getConnAtom(atom, conn);
          if (previousPath.includes(connectedAtom)) continue;
          nextIndexes.push(atomPaths.length);
          nextAtoms.push(connectedAtom);
          atomPaths.push({
            path: [...previousPath, connectedAtom],
            distance: sphere,
          });
        }
      }
    }
  }
  return allAtomsPaths;
}
