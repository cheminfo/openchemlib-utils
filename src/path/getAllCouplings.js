import { tagAtom } from '../util/tagAtom';
import { makeRacemic } from '../util/makeRacemic';

/**
 * Returns an array of all the different atom diaIDs that are connected
 * {OCL}
 * {OCL.Molecule} molecule
 * {object} [options={}]
 * {string} [options.fromLabel='H']
 * {string} [options.toLabel='H']
 * {number} [options.minLength=1]
 * {number} [options.maxLength=4]
 */

function getAllCouplings(OCL, molecule, options = {}) {
  const {
    fromLabel = 'H',
    toLabel = 'H',
    minLength = 1,
    maxLength = 4,
  } = options;
  let paths = molecule.getAllPaths({
    fromLabel,
    toLabel,
    minLength,
    maxLength,
  });

  const minSphereSize = 0;
  const maxSphereSize = 2;

  let fragment = new OCL.Molecule(0, 0);
  for (let path of paths) {
    path.info = [];
    for (let fromTo of path.fromTo) {
      let atoms = [];
      molecule.getPath(atoms, fromTo[0], fromTo[1], path.pathLength);
      let torsion;
      if (atoms.length === 4) {
        torsion = molecule.calculateTorsion(atoms);
      }
      path.info.push({
        atoms,
        torsion,
      });

      if (!path.code) {
        path.code = [];

        let tmpMolecule = molecule.getCompactCopy();
        makeRacemic(OCL, tmpMolecule);
        tagAtom(OCL, tmpMolecule, atoms[0]);
        tagAtom(OCL, tmpMolecule, atoms[atoms.length - 1]);

        let atomMask = new Array(tmpMolecule.getAllAtoms()).fill(false);
        let atomList = [];
        let max = 0;
        let min = 0;
        for (let sphere = 0; sphere <= maxSphereSize; sphere++) {
          if (max === 0) {
            for (let atom of atoms) {
              atomMask[atom] = true;
              atomList.push(atom);
              max++;
            }
          } else {
            let newMax = max;
            for (let i = min; i < max; i++) {
              let atom = atomList[i];
              for (let j = 0; j < tmpMolecule.getConnAtoms(atom); j++) {
                let connAtom = tmpMolecule.getConnAtom(atom, j);
                atomMask[connAtom] = true;
                atomList[newMax++] = connAtom;
              }
            }
            min = max;
            max = newMax;
          }
          tmpMolecule.copyMoleculeByAtoms(fragment, atomMask, true, null);
          if (sphere >= minSphereSize) {
            path.code.push(
              fragment.getCanonizedIDCode(
                OCL.Molecule.CANONIZER_ENCODE_ATOM_CUSTOM_LABELS,
              ),
            );
          }
        }
      }
    }
  }
  return paths;
}

module.exports = getAllCouplings;
