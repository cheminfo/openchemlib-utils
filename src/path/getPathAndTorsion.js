/**
 * Calculates the path between 2 atoms
 * @param {import('openchemlib').Molecule} molecule
 * @param {number} from - index of the first atom
 * @param {number} to - index of the end atom
 * @param {number} maxLength - maximal length of the path
 */
export function getPathAndTorsion(molecule, from, to, maxLength) {
  let originalAtoms = []; // path before renumbering
  molecule.getPath(originalAtoms, from, to, maxLength + 1);
  let torsion;
  if (originalAtoms.length === 4) {
    torsion = molecule.calculateTorsion(originalAtoms);
  }

  return {
    atoms: originalAtoms,
    from,
    to,
    torsion,
    length: originalAtoms.length - 1,
  };
}
