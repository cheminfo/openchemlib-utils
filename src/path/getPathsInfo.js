import { getHoseCodesForPath } from '../hose/getHoseCodesForPath';
import { getAtomsInfo } from '../util/getAtomsInfo';
import { getConnectivityMatrix } from '../util/getConnectivityMatrix';

let fragment;

/**
 *
 * @param {OCL.Molecule} molecule
 * @param {object} [options={}]
 * @param {string} [opions.fromLabel='H']
 * @param {string} [opions.toLabel='H']
 * @param {string} [opions.minLength=1]
 * @param {string} [opions.maxLength=4]

 */
export function getPathsInfo(molecule, options = {}) {
  const {
    fromLabel = 'H',
    toLabel = 'H',
    minLength = 1,
    maxLength = 4,
  } = options;

  const OCL = molecule.getOCL();

  if (!fragment) {
    fragment = new OCL.Molecule(0, 0);
  }

  let fromAtomicNumber = OCL.Molecule.getAtomicNoFromLabel(fromLabel);
  let toAtomicNumber = OCL.Molecule.getAtomicNoFromLabel(toLabel);

  // we need to find all the atoms 'fromLabel' and 'toLabel'
  let atomsInfo = getAtomsInfo(molecule);

  let pathLengthMatrix = getConnectivityMatrix(molecule, {
    pathLength: true,
  });

  for (let from = 0; from < molecule.getAllAtoms(); from++) {
    atomsInfo[from].paths = [];
    for (let to = 0; to < molecule.getAllAtoms(); to++) {
      if (from !== to) {
        if (molecule.getAtomicNo(from) === fromAtomicNumber) {
          if (molecule.getAtomicNo(to) === toAtomicNumber) {
            let pathLength = pathLengthMatrix[from][to];
            if (pathLength >= minLength && pathLength <= maxLength) {
              atomsInfo[from].paths.push(
                getHoseCodesForPath(
                  molecule,
                  from,
                  to,
                  pathLength,
                  atomsInfo[to].oclID,
                ),
              );
            }
          }
        }
      }
    }
  }

  return atomsInfo;
}
