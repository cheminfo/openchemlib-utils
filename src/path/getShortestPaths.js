/**
 * Get the shortest path between each pair of atoms in the molecule
 * @param {OCL.Molecule} molecule
 * @param {object} [options={}]
 * @param {string} [opions.fromLabel='H']
 * @param {string} [opions.toLabel='H']
 * @param {string} [opions.maxLength=4]
 * @returns {Array<Array>} A matrix containing on each cell (i,j) the shortest path from atom i to atom j
 */
export function getShortestPaths(molecule, options = {}) {
  const OCL = molecule.getOCL();
  const { fromLabel = '', toLabel = '', maxLength = 3 } = options;

  let fromAtomicNumber = OCL.Molecule.getAtomicNoFromLabel(fromLabel);
  let toAtomicNumber = OCL.Molecule.getAtomicNoFromLabel(toLabel);

  const nbAtoms = molecule.getAllAtoms();
  let allShortestPaths = new Array(nbAtoms);
  for (let i = 0; i < nbAtoms; i++) {
    allShortestPaths[i] = new Array(nbAtoms);
  }

  for (let from = 0; from < nbAtoms; from++) {
    allShortestPaths[from][from] = [from];
    for (let to = from + 1; to < nbAtoms; to++) {
      if (
        (fromAtomicNumber === 0 ||
          molecule.getAtomicNo(from) === fromAtomicNumber) &&
        (toAtomicNumber === 0 || molecule.getAtomicNo(to) === toAtomicNumber)
      ) {
        let path = [];
        molecule.getPath(path, from, to, maxLength);
        if (path.length) {
          allShortestPaths[from][to] = path.slice();
          allShortestPaths[to][from] = path.reverse();
        } else {
          allShortestPaths[from][to] = null;
          allShortestPaths[to][from] = null;
        }
      } else {
        allShortestPaths[from][to] = null;
        allShortestPaths[to][from] = null;
      }
    }
  }

  return allShortestPaths;
}
