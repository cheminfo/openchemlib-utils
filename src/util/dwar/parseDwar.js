import { ensureString } from 'ensure-string';

import { getParts } from './utils/getParts.js';
import { parseColumnbProperties } from './utils/parseColumnbProperties.js';
import { parseData } from './utils/parseData.js';

/**
 * Convert a DataWarrior database into a JSON object
 * @param {string} text
 * @returns
 */
export function parseDwar(text) {
  text = ensureString(text);
  const parts = getParts(text);
  improveParts(parts);
  return parts;
}

function improveParts(parts) {
  for (const key in parts) {
    switch (key) {
      case 'columnProperties':
        parts[key] = parseColumnbProperties(parts[key]);
        break;
      case 'data':
        break;
      default:
        parts[key] = parseDefault(parts[key]);
    }
  }
  const data = parseData(parts.data, {
    columnProperties: parts.columnProperties,
  });
  parts.data = data.entries;
  parts.rawData = data.rawEntries;
}

function parseDefault(lines) {
  const result = {};
  for (const line of lines) {
    const [key, value] = line.slice(1, -1).split('=');
    result[key] = value.slice(1, -1);
  }
  return result;
}
