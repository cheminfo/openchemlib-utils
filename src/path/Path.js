import { getPathsInfo } from './getPathsInfo';

export const Path = {
  getPathsInfo,
};
