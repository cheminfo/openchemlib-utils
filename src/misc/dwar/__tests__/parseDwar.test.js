import { readFileSync } from 'fs';
import { join } from 'path';

import { parseDwar } from '../parseDwar.js';

test('parseDwar', () => {
  const text = readFileSync(join(__dirname, 'data/reactions.dwar'), 'utf8');
  const reactions = parseDwar(text);
  const data = reactions.data;
  expect(data).toHaveLength(3);
  expect(reactions).toMatchSnapshot();
});
