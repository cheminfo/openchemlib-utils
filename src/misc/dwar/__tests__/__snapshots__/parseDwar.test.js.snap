// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseDwar 1`] = `
{
  "columnProperties": {
    "atomMapping": {
      "parent": "rxnCode",
      "specialType": "atomMapping",
    },
    "idcoordinates2D": {
      "parent": "rxnCode",
      "specialType": "idcoordinates2D",
    },
    "productFFP": {
      "parent": "rxnCode",
      "reactionPart": "products",
      "specialType": "FragFp",
      "version": "1.2.1",
    },
    "reactantFFP": {
      "parent": "rxnCode",
      "reactionPart": "reactants",
      "specialType": "FragFp",
      "version": "1.2.1",
    },
    "reactionFP": {
      "parent": "rxnCode",
      "specialType": "RxnFP",
      "version": "1.0.0",
    },
    "rxnCode": {
      "isFragment": "true",
      "related": {
        "FragFp": "productFFP",
        "RxnFP": "reactionFP",
        "atomMapping": "atomMapping",
        "idcoordinates2D": "idcoordinates2D",
      },
      "specialType": "rxncode",
    },
  },
  "data": [
    {
      "Label": "OH by SH",
      "rxnCode": "eFHBLGtV!eFB@HcA}E\`#aP aP#!R@AL]\\mp !R@AL]Nmp",
    },
    {
      "Label": "SH to SMe",
      "rxnCode": "eFB@HcAa@!eMB@HchH#S\\ S\\@#!R@FL]bgp !R@FL?XqwZ_@",
    },
    {
      "Label": "SH to OH",
      "rxnCode": "eFB@HcA}E\`!eFHBLGtV#aP aP#!R@FL]Rgp !R@FL]vgp",
    },
  ],
  "datawarriorFileinfo": {
    "created": "1674563264760",
    "rowcount": "6",
    "version": "3.3",
  },
  "datawarriorProperties": {
    "columnFilter_Table": "",
    "columnWidth_Table_Label": "228",
    "columnWidth_Table_rxnCode": "393",
    "detailView": "height[Data",
    "filter0": "#string#	Label",
    "filter1": "#reaction#	rxnCode",
    "filter2": "#retron#	rxnCode",
    "headerLines_Table": "2",
    "mainSplitting": "0.72655",
    "mainView": "Table",
    "mainViewCount": "1",
    "mainViewDockInfo0": "root",
    "mainViewName0": "Table",
    "mainViewType0": "tableView",
    "rightSplitting": "0.65678",
    "rowHeight_Table": "80",
  },
  "hitlistData": {
    "hitlistData": "D",
    "hitlistName": "<selection>",
  },
  "rawData": [
    {
      "Label": "OH by SH",
      "atomMapping": "aP aP",
      "idcoordinates2D": "!R@AL]\\mp !R@AL]Nmp",
      "productFFP": "//,8///$",
      "reactantFFP": "40V/////,10",
      "reactionFP": "!F!4$840F!1F&2V#F4+F/F01#12V42010k#G$8102!V//////0",
      "rxnCode": "eFHBLGtV!eFB@HcA}E\`",
    },
    {
      "Label": "SH to SMe",
      "atomMapping": "S\\ S\\@",
      "idcoordinates2D": "!R@FL]bgp !R@FL?XqwZ_@",
      "productFFP": "2//+8///$",
      "reactantFFP": "//,8///$",
      "reactionFP": "2!52&4$1020V&VJ&V$V!4!1%8#FV1#22V8(1%8//(1//%8/(",
      "rxnCode": "eFB@HcAa@!eMB@HchH",
    },
    {
      "Label": "SH to OH",
      "atomMapping": "aP aP",
      "idcoordinates2D": "!R@FL]Rgp !R@FL]vgp",
      "productFFP": "40V/////,10",
      "reactantFFP": "//,8///$",
      "reactionFP": "!18$4(1%4(4$3!480F$F#24$F!2#2V02020F08201#81#F//////!",
      "rxnCode": "eFB@HcA}E\`!eFHBLGtV",
    },
  ],
}
`;
